pkg/operator/client/delete.go

```diff
@@ -0,0 +1,392 @@
+			if err != nil {
+				return errors.Wrapf(err, "failed to list resources to delete in namespace %s", namespace)
+			}
+			resourcesToDelete = append(resourcesToDelete, rs...)
+		}
+
+		if len(resourcesToDelete) == 0 {
+			logger.Infof("Successfully cleared all resources for app %s\n", appSlug)
+			return nil
+		}
+
+		resourcesToDelete = sortResourcesForDeletion(resourcesToDelete)
+
+		for _, r := range resourcesToDelete {
+			if r.Unstructured.GetDeletionTimestamp() != nil {
+				logger.Infof("Pending deletion %s/%s/%s/%s/%s", r.Unstructured.GetNamespace(), r.GVR.Group, r.GVR.Version, r.GVR.Resource, r.Unstructured.GetName())
+				continue
+			}
+
+			logger.Infof("Deleting %s/%s/%s/%s/%s", r.Unstructured.GetNamespace(), r.GVR.Group, r.GVR.Version, r.GVR.Resource, r.Unstructured.GetName())
+
+			if err := dyn.Resource(r.GVR).Namespace(r.Unstructured.GetNamespace()).Delete(context.TODO(), r.Unstructured.GetName(), metav1.DeleteOptions{}); err != nil {
+				logger.Errorf("Resource %s/%s/%s/%s/%s could not be deleted: %v", r.Unstructured.GetNamespace(), r.GVR.Group, r.GVR.Version, r.GVR.Resource, r.Unstructured.GetName(), err)
+			}
+		}
+
+		time.Sleep(sleepTime)
+	}
+
+	return fmt.Errorf("failed to clear all resources for app %s", appSlug)
+}
+
+func (c *Client) deletePVCs(appLabelSelector *metav1.LabelSelector, appslug string) error {
+	clientset, err := k8sutil.GetClientset()
+	if err != nil {
+		return errors.Wrap(err, "failed to get k8s clientset")
+	}
+
+	if appLabelSelector == nil {
+		appLabelSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{},
+		}
+	}
+	appLabelSelector.MatchLabels["kots.io/app-slug"] = appslug
+
+	podsList, err := clientset.CoreV1().Pods(c.TargetNamespace).List(context.TODO(), metav1.ListOptions{
+		LabelSelector: getLabelSelector(appLabelSelector),
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to get list of app pods")
+	}
+
+	pvcs := make([]string, 0)
+	for _, pod := range podsList.Items {
+		for _, v := range pod.Spec.Volumes {
+			if v.PersistentVolumeClaim != nil {
+				pvcs = append(pvcs, v.PersistentVolumeClaim.ClaimName)
+			}
+		}
+	}
+
+	if len(pvcs) == 0 {
+		logger.Infof("no pvcs to delete in %s for pods that match %s", c.TargetNamespace, getLabelSelector(appLabelSelector))
+		return nil
+	}
+	logger.Infof("deleting %d pvcs in %s for pods that match %s", len(pvcs), c.TargetNamespace, getLabelSelector(appLabelSelector))
+
+	for _, pvc := range pvcs {
+		grace := int64(0)
+		policy := metav1.DeletePropagationBackground
+		opts := metav1.DeleteOptions{
+			GracePeriodSeconds: &grace,
+			PropagationPolicy:  &policy,
+		}
+		logger.Infof("deleting pvc: %s", pvc)
+		err := clientset.CoreV1().PersistentVolumeClaims(c.TargetNamespace).Delete(context.TODO(), pvc, opts)
+		if err != nil && !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrapf(err, "failed to delete pvc %s", pvc)
+		}
+	}
+
+	return nil
+}
+
+func getLabelSelector(appLabelSelector *metav1.LabelSelector) string {
```

Comment:
i think this function might already exist in `"k8s.io/apimachinery/pkg/labels"` as `SelectorFromSet(ls labels.Set) labels.Selector`?  we use that in a bunch of places